### DISCUSSION TITLE
Project Future

### DISCUSSION THREAD

**Discussion by ds5678** [AssetRipper/AssetRipper#953] [Announcements]
### Describe the issue

I've worked on AssetRipper for a year and spent thousands of hours in its development. In order to continue sustainable development on AssetRipper, I am moving towards relying on a freemium model to fund it. This will allow me to maximize its potential without worrying as much about expenses, such as rent or internet. If AssetRipper does not generate sufficient funding, I will be forced to slow, pause, or abandon its development. 

Currently, I plan for this change to happen when I overhaul the GUI and replace it with an Electron solution. I can't say for certain when that will happen, since I'm still bad at making accurate time estimations. I consider this the next major milestone for the project. Significant portions of my focus will be dedicated to making it a reality. I will probably dub this version either `0.4.0.0` or `1.0.0.0`.

## What this means for users

AssetRipper will still be usable by anyone. However, there are some things of note:

* It will still be a desktop application.
* It will still not require internet.
* It will be NativeAOT compiled for performance, file size, and security.
* There will be free versions available on GitHub and premium versions available on Patreon.
  * Premium versions will have extra features.
  * A free feature will only become premium if substantial improvements are made to it.
    * For example, static mesh separation is experimental and mostly broken. If I put lots of time into making it fully usable, it will be premium.
  * I am considering offering some benefit to past donators as a token of my thanks for their past generosity.
    * One idea was temporary access to the premium versions without a subscription.
  * The premium tier on Patreon (probably $5) will give access to all current and past premium versions.
  * Once downloaded, premium versions last forever. Subscription is only required to get newer versions.
  * I will probably offer Patreon polls so that subscribers can vote on features to be implemented, for both the free and premium versions.
  * Some features may be premium for a set time before being released as free and open source.
    * For example, unreal support would likely be this way if I decide to implement it.

## What this means for future contributors

I always appreciate people's willingness to help develop a project. Most of AssetRipper will stay open-source and open to contributions. However, all contributors must agree to AssetRipper's contribution license agreement in order for their code changes to be merged. This ensures that I remain in control of the codebase and able to distribute AssetRipper under a non-GPL license, if desired.

## My Appreciation

I hope you understand why this needs to happen. I know this may be disappointing or infuriating to some of you, and I am sorry. I wish AssetRipper could remain fully open-source and worked on by everyone. I wish I didn't have to worry about food or rent. I've spent so much time working on AssetRipper so that others could have an enjoyable experience using it. I don't want to have to stop due to financial burden. If this succeeds, I will be so grateful. You mean a lot to me.

## Updates

* August 1, 2023

**Comment by Grape-Boy** [comment]
dam maybe you could also get a patreon account

**Comment by ds5678** [comment]
> dam maybe you could also get a patreon account

I have multiple sponsorship accounts:
* [GitHub Sponsors](https://github.com/sponsors/ds5678)
* [Patreon](https://patreon.com/ds5678)
* [Paypal](https://paypal.me/ds5678)

**Comment by Grape-Boy** [comment]
nice

**Comment by cameron-macy** [comment]
"Showing ads (or maybe paying for ad removal) will improve export quality", How would you decrease export quality, limited features? And if so, which ones?
How much do you estimate its going to cost?
what are you going to do if some of the largest contributors decide to not cooperate with your cla or an important pull like "Panthr75:fix-method-stripping"?

**Comment by ds5678** [comment]
> "Showing ads (or maybe paying for ad removal) will improve export quality", How would you decrease export quality, limited features? And if so, which ones?

You're thinking about this in the wrong light. It's not about decreasing the quality. I'll implement some highly desired improvements as allow-ads-only. The ad-blocked version of AssetRipper isn't being crippled. I'm rewarding users who allow ads.

> How much do you estimate its going to cost?

Paying for ad removal will not be available at launch and is not guaranteed to happen. If I add that option at some point, I'll thoroughly discuss it with the community beforehand.

> what are you going to do if some of the largest contributors decide to not cooperate with your cla

Unfortunately, if any existing contributors do not sign the CLA, I will have to remove the contributions of those people from the codebase. Thankfully, most of the contributors have either already signed or expressed their intention to sign.

> an important pull like "Panthr75:fix-method-stripping"?

I greatly appreciate when users decide to contribute their time into the project. I consider it a show of thanks for all the work I've done on the project. Although some pull requests may be more important than others, I will not bend over backwards for any pull request, especially since contributors often do not stay around to maintain their changes and fix resulting bugs. That falls to me and takes just as much work as the initial implementation.

**Comment by cameron-macy** [comment]
Thank you.

**Comment by bb010g** [comment]
Will patches to open-source AssetRipper components that reproduce existing closed-source AssetRipper features be accepted?

**Comment by ds5678** [comment]
> Will patches to open-source AssetRipper components that reproduce existing closed-source AssetRipper features be accepted?

Such action is an ill-advised waste of development resources. The idea comes across as antagonistic, especially if the feature was "reproduced" via attempted reverse engineering of AssetRipper binaries. In addition, unless the quality of the reproduction exceeds that of the original, there's no reason for me to merge the reproduction instead of simply opensourcing the code for the original.

In answers to related questions on the AssetRipper Discord server, I've expressed a willingness to slowly opensource closed source features. This would likely depend on a variety of factors, such as:
* How long the feature has been implemented
* The amount of contributor interest in working on the feature
* The risk of competitors copying the feature
* The amount of maintenance required for the feature
* The completeness of the feature

**Comment by dukemagus** [comment]
Hooking the program to a regular browser webUI is a security risk, that's for sure.
It's easier to make a standalone app and some ad space on the corners of the editor.

You'll also need to curate your ad provider very carefully and constantly unless you want your program to be associated with viruses, malware, phishing, illegal gambling, pornography, prostitution (illegal in some countries), scams, etc.

And GPL is an infectious license. To close the source, you need to drop any external contribution accepted so far, since all other contributors added their code under GPL3.

I ask you to consider other more clear-cut solutions rather than ads (also, the target audience of the app doesn't overlap with most of the run of the mill ads i see on adsense or facebook).

- **Add a commercial option while keeping the source availiable:**

  - [Aseprite](https://github.com/aseprite/aseprite) and [Armorpaint](https://github.com/armory3d/armorpaint) are sold commercially and also have the source availiable. The catch is that **if you want to use it for free, you have to compile it yourself**. They sell the convenience of a compiled and documented software with customer support on [Steam](https://store.steampowered.com/app/431730/Aseprite/) (Aseprite) and [Gumroad](https://armory.gumroad.com/l/armorpaint) (Armorpaint) Since both are still being developed, it should have been sucessful to some extent. Even with a "free" port of an older version (Libresprite), Aseprite still sells decently on Steam.

- **Put highly requested features on a bounty site** (ex: https://bountysource.com). 
  - That way, the more enthusiast members of the community can contribute and everyone can benefit from the work done on the issue.

- **Make the donation links more visible and embed them to the GUI:**
  - As it stands, the links on the readme have nothing to make them stand out. Other project add a lot of clarification and transparecny information, eye catching and focus on their donation pages, for example: https://pi-hole.net/donate/ or https://wiki.freecadweb.org/Donate

- **Add other payment options more commonly associated with software development.**
  - [Liberapay](https://liberapay.com), [Open Collective](https://opencollective.com/opensource) and the previously mentioned Bountysource are much more popular and more frequently linked on software development and technology forums due to their focus being specifically on open source software

Wish you the best of luck.

**Comment by ds5678** [comment]
@dukemagus I see you've put a lot of thought into this.

> Hooking the program to a regular browser webUI is a security risk, that's for sure.

As far as I understand, this is no more risky than opening a normal websocket to the internet since standalone applications can access most files on your computer.

> It's easier to make a standalone app and some ad space on the corners of the editor.

All the ad providers I've found are only willing to provide ads for websites and mobile apps.

> You'll also need to curate your ad provider very carefully and constantly unless you want your program to be associated with viruses, malware, phishing, illegal gambling, pornography, prostitution (illegal in some countries), scams, etc.

Of course. I want people to feel safe while using AssetRipper.

> And GPL is an infectious license. To close the source, you need to drop any external contribution accepted so far, since all other contributors added their code under GPL3.

All contributor licensing issues were handled in #375.

> Add a commercial option while keeping the source availiable:

I'm skeptical of this solution because compiling C# is way easier than C++, and I don't want to have to worry about secondary distribution.

> Put highly requested features on a bounty site (ex: https://bountysource.com/).

I'm not familiar with this site, but I briefly looked over it. I assume you mean me posting an issue and other people offering money for the issue to be resolved. I'm not opposed to the idea. It would give development agency to the more generous users, which is good. The only thing I don't like about it is that it seems to reward feature creation and ignores feature maintenance, which is more boring and takes just as much work, unless I neglect to fix bugs until someone is willing to pay. Neglecting bugs for that reason would feel very scummy to me and hurt the trust I have with my users.

> Make the donation links more visible and embed them to the GUI:
> Add other payment options more commonly associated with software development.

Other than a one-time donation in exchange for an implemented feature, I've only ever received about $50 at most from AssetRipper users, so I don't have a lot of faith in the donation system. However, every bit helps, and I don't see any reason why I shouldn't expand the visibility and variety of donation options.

**Comment by bb010g** [comment]
@ds5678

> > Add a commercial option while keeping the source availiable:
> 
> I'm skeptical of this solution because compiling C# is way easier than C++, and I don't want to have to worry about secondary distribution.

I'm also a fan of Aesprite's business model. Something they do effectively is make it clear that the product is paid. The source happens to be available too, but you *should* buy a license if you're building it yourself. As for how difficult it is to build, I think if you make the non-GitHub website the place people start at (which advertises buying the license) and put up stable and nightly builds for those with a license (with a good updating solution, like Sparkle), it won't be much of an issue. People will pay for the convenience of keeping the application up-to-date without having to redownload manually, let alone recompile manually. You can also make it clear, in the issue tracker, that issues and pull requests from license-holders are prioritized. Nobody's going to fault you for supporting your paying customers.

**Comment by ds5678** [comment]
@bb010g if I were to go along with this idea, what should I charge and how often?

**Comment by bb010g** [comment]
@ds5678 Your pricing depends on what you think your product is worth, how many customers you have, how much those customers think your product is work, how much prospective customers think your product is worth, and a lot of other stuff. You're almost certainly going to be able to come up with better numbers here than me. I can give you some points to start from, though, from software I've paid for. Aseprite sells a perpetual license to Aseprite 1.x releases for US$20 (<https://www.aseprite.org/faq/#what-do-i-get-when-i-buy-aseprite>). Sublime HQ sells a perpetual personal (commercial ok, businesses can't own this one) license to Sublime Merge, with 3 years of updates, for US$100, and they sell a subscription business (the business can own this one) license to Sublime Merge, for 1 year with updates, for US$75/year (<https://www.sublimehq.com/store/merge>). Cryptic Apps sells a perpetual personal (commercial ok) license to Hopper Disassembler, with 1 year of updates, for US$100 (<http://www.hopperapp.com/>). Vector 35 sells a personal ([non-commercial](https://binary.ninja/faq/#commercial-use) (making less than 1/5 of the cost of a license using the software is allowed)) license to Binary Ninja, with 1 year of updates, for US$300 (not counting the 75% student/education discount), they sell a commercial license to Binary Ninja, with 1 year of updates, for US$1200, and they also sell enterprise licenses to Binary Ninja (<https://binary.ninja/purchase/>). (Note that their FAQ contains the entry [*Can you write a sole source justification for my purchasing department?*](https://binary.ninja/faq/#sole-source).)

I'd probably look take influence from the Sublime Merge and Binary Ninja licensing. If you have a personal license, attach it to a specific person and don't let it be owned by a business. It's useful for a business to own a license that they can pass around between team members. If you go that route, you probably want to not make the business license perpetual, like with Sublime Merge. If you make the personal license non-commercial, you probably want to use a similar "de minimis" standard to what Binary Ninja uses (this covers cases like small YouTube channels, where money is technically being made off of the licensed software but it's pretty insignificant and the PR is worth it). Remember to market that a license buys you meaningful support, updates, and ensures the software will continue to be around in a couple of years. If you expect game studios will be buying only a one or a few licenses, it might be good to price what they'd be buying higher. Perhaps you can set a concrete upper cap where you feel is reasonable for the full worth of the software, and then if that'd be too expensive for indie devs that you'd expect to be using it, make provisions for it to be cheaper just for them. If you have commercial users, now is a great time to reach out to them and see how they've used your software, and how much they feel it's been worth to them / what they'd be willing to pay for the functionality they've experienced so far.

**Comment by zone-a-1** [comment]
You could use something like Bing Ads and keep the C# standalone app: https://docs.microsoft.com/en-us/advertising/guides/get-started-csharp?view=bingads-13 https://docs.microsoft.com/en-us/advertising/guides/walkthrough-desktop-application-csharp?view=bingads-13

**Comment by heinermann** [comment]
> [Aseprite](https://github.com/aseprite/aseprite) and [Armorpaint](https://github.com/armory3d/armorpaint) are sold commercially and also have the source availiable. The catch is that if you want to use it for free, you have to compile it yourself. They sell the convenience of a compiled and documented software with customer support on [Steam](https://store.steampowered.com/app/431730/Aseprite/) (Aseprite) and [Gumroad](https://armory.gumroad.com/l/armorpaint) (Armorpaint) Since both are still being developed, it should have been sucessful to some extent. Even with a "free" port of an older version (Libresprite), Aseprite still sells decently on Steam.

I think the major downside to this is some modding communities are beginning to rely on AssetRipper to make things easier, and putting all releases behind a paywall could stifle modding for those communities. Even if it incites some people to purchase, others will be dissuaded which will prevent them from modding effectively (barrier to entry). So if this model is adopted, I'd like for it to be balanced with this drawback in some way, for example by offering older releases for free after a certain time period or keeping the default and unity-compatible export options free but other options paid. Though I also understand if this isn't possible.

**Comment by sparr** [comment]
I thought it worth chiming in as a command line user, who intends to release a command line program that calls AssetRipper... Whatever is available open source will remain useful to me/us, and anything happening in the GUI or web version is outside our concern.

**Comment by dukemagus** [comment]
Delaying the repo release so the paying customers always have the most
updated and stable product is also an option

Plenty of people sell games and tools on itch.io and Patreon while at the
same time sharing a link to the previous version on the page (or waiting
one or two weeks to share the version regardless of how new it is)

Em qui., 16 de jun. de 2022 13:06, bb010g ***@***.***>
escreveu:

**Comment by ghost** [comment]
"Once GPL, Always GPL". You literally can't switch

**Comment by ghost** [comment]
> You could use something like Bing Ads and keep the C# standalone app: https://docs.microsoft.com/en-us/advertising/guides/get-started-csharp?view=bingads-13 https://docs.microsoft.com/en-us/advertising/guides/walkthrough-desktop-application-csharp?view=bingads-13

Bing sux lmao

**Comment by ds5678** [comment]
> "Once GPL, Always GPL". You literally can't switch

That is not how licensing works. All the contributors signed a CLA, and the CLA allows me to do whatever I want. There's no external GPL code used in AssetRipper, so I could make a license change if I wanted.

**Comment by ghost** [comment]
Then make sure that instead of going proprietary/closed-source you get
yourself a patreon, which would be better. If you don’t comply, I will
rebuild uTinyRipper for Mac. And I will do it way better than you. So will
you comply?

**Comment by bb010g** [comment]
> Then make sure that instead of going proprietary/closed-source you get yourself a patreon, which would be better.

If you read the thread, you'll see that they already have a Patreon account.

> If you don’t comply, I will rebuild uTinyRipper for Mac. And I will do it way better than you. So will you comply?

Let's not be needlessly petty here.

**Comment by ghost** [comment]
1. The thread is huge.
2. OK!

**Comment by DragoonX6** [comment]
A possible option is to release new features to patreon supporters, so that people who want the new features will have to support you on patreon. Then after a certain amount of time (say 6 months) you can open source the oldest feature(s).  
I think this is a model that works as I've seen other developers successfully do this before and results in little to no friction with the community.

Personally I don't want to see this move to the browser, nor include ads (for many reasons), as that would ruin the usage of the software for me. I also think you're overestimating the revenue you would get from this as most people would open the website, dump, and then leave. I think it's better to make it just paid, or slow down development and push for more sponsor money.

**Comment by ghost** [comment]
Great idea! 10/10

**Comment by ds5678** [comment]
Hi guys, just a quick status update:

I've been job searching, so things might be slower than before, but not non-existent. Sometime in the near future, I'll be doing a `0.3.0.0` release (which is not related to any mentions of `0.3.0.0` above) with a variety of big improvements.

Although I may try out the ad-based model *at some point*, there is not currently any work being done towards that. My current plans are to be more explicit in the user interface about the need for donation support because many users are not even *aware* that AssetRipper has donation options.

**Comment by ghost** [comment]
> Hi guys, just a quick status update:
> 
> I've been job searching, so things might be slower than before, but not non-existent. Sometime in the near future, I'll be doing a `0.3.0.0` release (which is not related to any mentions of `0.3.0.0` above) with a variety of big improvements.
> 
> Although I may try out the ad-based model _at some point_, there is not currently any work being done towards that. My current plans are to be more explicit in the user interface about the need for donation support because many users are not even _aware_ that AssetRipper has donation options.

Please don’t introduce nag screens.

**Comment by ds5678** [comment]
@h57758 as far as I know, you are not a financial supporter of AssetRipper. This is unfortunate because you have such strong opinions on its future development. If you, and the other non-supporters, want development move in directions you desire, I need stronger financial encouragement to do so.

**Comment by ghost** [comment]
I might try to help, but the only problem is that I don’t have much money :(

**Comment by DragoonX6** [comment]
> Although I may try out the ad-based model _at some point_, there is not currently any work being done towards that. My current plans are to be more explicit in the user interface about the need for donation support because many users are not even _aware_ that AssetRipper has donation options.

I think this is a good idea, and optionally you could let the user know of your alternate ad plans as extra incentive to get people to support the development of AssetRipper.

**Comment by ghost** [comment]
No please, not ads! I’d rather have a nag screen every minute than ads. If
you add ads, I’m switching back to uTinyRipper

**Comment by ds5678** [comment]
>  I don’t have much money

Everyone is short on funds. I am no exception to this, and yet I support the creators I benefit from. People can't afford to work for free, and it's ungrateful to expect them to.

> No please, not ads! I’d rather have a nag screen every minute than ads. 

If people won't willingly support my work, I have to find other alternatives. Some of these alternatives (such as ads) might be unpopular. I am okay with that.

> If you add ads, I’m switching back to uTinyRipper

You are welcome to do so. There have been 2000 AssetRipper commits since I forked uTinyRipper, so you would be missing out on a lot.

**Comment by ghost** [comment]
3. Stop throwing propaganda around!

**Comment by iruzer** [comment]
I have plan to donate with patreon if you often update news or there is exclusive feature for patreon donate.

**Comment by ds5678** [comment]
> I have plan to donate with patreon if you often update news or there is exclusive feature for patreon donate.

I have been planning to expand my Patreon presense. This may initially take the form of development polls and special news posts. However, I've also been considering a Freemium model where I implement some extra features and make a Premium release for subscribers. I'm open to suggestions on what features might go into a Premium release.

**Comment by ModischFabrications** [comment]
Just came through looking for a modern version of uTinyRipper. I can't/won't argue about monetization, but only on usability.  That said, take my comments with a grain of salt, I am still researching if this is the right tool for the job and am absolutely clueless on the internals. 

The hybrid solution of a website with additional executable worker to download feels a bit weird. I would expect either A: a webversion to work without any additional downloads or B: an executable that offers a GUI or CLI and works "as-is", without any browser. 
I agree that embedded ads inside an executable feel more serious than ads on a website that controls a closed-source application that works in the background without any observability.

**Comment by paomeng-game** [comment]
Thank you for your endless contributions to AssetRipper, which I use to respawn a great Unity game I love most during my childhood.

**Comment by paomeng-game** [comment]
Hey , ds5678 , I just notices that you enable Sponsor. I think you are lack of funding. What I want to be is your Sponsor , Although I am not a rich man,  but the problem is that I am a Chinese, It is so hard for me to sign up a account on some paying website such as paypel. We usually fond of Alipay or Wechat pay. I am very sad about this. If you can upload your QR code with Alipay and Wechat Pay, I may help you a little. Thank you very much!

**Comment by Lachee** [comment]
Ill post here what I replied to you before 

> Users will access AssetRipper via a website. Consequently, it requires constant internet access to use.

This is dumb and is entirely unnecessary for a tool. If you want to make a premium website sure, but forcing this tool to use it I disagree.

> Users may be required to download a "worker" application that handles most of the heavy lifting for AssetRipper.

So ripping the guts of the app to put in ... an app.... that is just controlled via a web ui?

> The worker application will be native compiled and maybe obfuscated before compilation.

Why!? This code is GPL. This is the worst part of this proposal. Its a gross mistrust and just obfuscating it will remove any kind of trust.

> Currently, AssetRipper releases are distributed with the GPL3 license. In the future, they'll be distributed with a different license reflecting its closed-source nature. 

I use your tpp; because the alternative is a shady program that is full of malware. Close-Sourcing it will make you as trust worthy as that shady application.

> Showing ads

No

> Using an adblock will limit you to free features.

My data, my net, my screen, MY COMPUTER. I chose what I want to show on it, if I run adblock that is my choice and my right to not download garbage. 

> automated bug reporting

This needs to be a opt-in (or out). 

>  I am investigating the legal requirements for doing this safely.

what?


**Edit:**

Apparently i am misinformed from the discussion and this is now outdated information, I apologise. However, I will retain my reply to the _old_ plan regardless as my opinion to the original discussion points as to the monitisation scheme.

**Comment by CodeEngieer** [comment]
Hello author, thank you very much for open source this project, I use this tool to learn coding skills at work and successfully work in this field for many years, I would like to ask whether il2cpp can crack the corresponding source code in the future, for example, il2cppdumper can only decompilate dll files. The corresponding source code only has attributes and method names, and you can't see the specific implementation process of the method. And the premium version of sponsors can I have the problem I'm having now?

**Reply by ds5678** [reply]
That is not currently planned for implementation.

**Comment by BlueFalconHD** [comment]
@ds5678 Have you considered allowing heavy and or large contributors to get access to premium versions for free? Or giving them access to the premium version repository, so they can continue to contribute? This would serve as an incentive to do good work on open source software and could take some of the pressure away from you.

**Reply by ds5678** [reply]
> Have you considered allowing heavy and or large contributors to get access to premium versions for free?

I'd gladly reward these contributors in that way.

> Or giving them access to the premium version repository, so they can continue to contribute?

I would be hesitant to give anyone access. I have not encountered sufficient justification to do so.

**Comment by BlueFalconHD** [comment]
I support the need for funding and recognize the tremendous effort put into AssetRipper's development. However, I believe the risks of a freemium model outweigh the rewards. Closing the source code limits learning opportunities and increases the risk of impersonation attacks that could spread malware, leading to a loss of trust among users. Balancing sustainable development with maintaining an open-source environment is crucial for the project's long-term success and community support.

**Reply by ds5678** [reply]
With all due respect, the freemium model has saved the project.

**Reply by BlueFalconHD** [reply]
Understood, and I am glad you are able to financially support yourself. However, some of the stuff Assetripper does can be learned from. Maybe releases of premium from 2+ years prior could be merged with the main source, so those who are financially able to support you could still access newer features, but not the state of the art. Also, having premium could be a means to get immediate or very fast support or help.
